keps/sig-node/24-apparmor-ga/README.md

```diff
@@ -0,0 +1,682 @@
+| -------------------------------------------------------------------------------------------------- | -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| 1) Using custom or `runtime/default` profile when container runtime does not support AppArmor.     | Pod created          | The outcome is container runtime dependent. In this scenario containers may 1) fail to start or 2) run normally without having its policies enforced. |
+| 2) Using custom or `runtime/default` profile that restricts actions a container is trying to make. | Pod created          | The outcome is workload and AppArmor dependent. In this scenario containers may 1) fail to start, 2) misbehave or 3) log violations.                  |
+| 3) Using custom profile that does not exist on the node.                                           | Pod created          | Containers fail to start. Retry respecting RestartPolicy and back-off delay.                                                                          |
+| 4) Using an unsupported runtime profile (i.e. `runtime/default-audit`).                            | Pod **not** created. | N/A                                                                                                                                                   |
+| 5) Using localhost profile with invalid name                                                       | Pod **not** created. | N/A                                                                                                                                                   |
+| 6) AppArmor is disabled by the host or the build                                                   | Pod **not** created. | Kubelet puts Pod in blocked state.                                                                                                                    |
+
+Scenario 2 is the expected behavior of using AppArmor and it is included here
+for completeness.
+
+Scenario 5 represents the case of failing the existing validation, which is
+defined at [Pod API](#pod-api).
+
+### Version Skew Strategy
+
+Because the API is currently represented as (mutable) annotations, care must be
+taken for migrating to the API fields. The cases to consider are Pod create and
+update.
+
+All API skew is resolved in the API server. New Kubelets will only use the
+AppArmor values specified in the fields, and ignore the annotations.
+
+#### Pod Creation
+
+If no AppArmor annotations or fields are specified, no action is necessary.
+
+If _only_ AppArmor fields are specified, add the corresponding annotations. This
+ensures that the fields are enforced even if the node version trails the API
+version (see [Version Skew Strategy](##version-skew-strategy))
+
+If _only_ AppArmor annotations are specified, copy the values into the
+corresponding fields. This ensures that existing applications continue to
```

Comment:
an empty string annotation value or a `localhost/` annotation value is allowed (https://github.com/kubernetes/kubernetes/blob/deebe6619a1b4955b0cd109a8f05b1298a260310/pkg/security/apparmor/validate.go#L138-L142). what would happen with those?

Reply:
CRI-O and containerd:
- apply the runtime default profile if the name is empty
- strip the `localhost/` prefix

Reply:
add descriptions of those edge cases and what field values we will translate those to (and make sure the impl PR tests those)

Reply:
Done, also added a note about the e2e tests.